{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Questions on Data Sources\n",
    "\n",
    "\n",
    "## CGDS\n",
    "\n",
    "- Data sets available (http://www.cbioportal.org/data_sets.jsp)\n",
    "    - What are RNA-Seq V2 and miRNA datasets?\n",
    "- Is there any easy way to tell which cell lines were part of which studies?  Or intersections of genes?\n",
    "\n",
    "\n",
    "## GDSC\n",
    "\n",
    "- **Copy Number**\n",
    "    - Downloaded spreadsheet: ~/repos/mgds/python/notebook/data_aggregation/raw/gdsc/v1/downloads/copy-number.xlsx\n",
    "    - What are the different types of copy number data (see http://www.cancerrxgene.org/downloads)?\n",
    "    - What is the different between the raw Affymetrix data and the preprocessed?\n",
    "    - What are the RACs data sets?\n",
    "    - [\"RACS in cell lines\"](ftp://ftp.sanger.ac.uk/pub/project/cancerrxgene/releases/release-6.0/RACS_in_cell_lines.xlsx) returns a spreadsheet that looks like this:\n",
    "    \n",
    "            Sample\tCOSMIC_ID\tTissue identifier 1\tTissue identifier 2\tCancer Type\tAlteration Type\tRegion identifier\n",
    "            201T\t1287381\tlung_NSCLC\tlung_NSCLC_adenocarcinoma\tLUAD\tDeletion\tcnaLUAD1 (CDKN2A)\n",
    "            201T\t1287381\tlung_NSCLC\tlung_NSCLC_adenocarcinoma\tLUAD\tAmplification\tcnaLUAD3 (TERT)\n",
    "            201T\t1287381\tlung_NSCLC\tlung_NSCLC_adenocarcinoma\tLUAD\tAmplification\tcnaLUAD17 (KRAS)\n",
    "            201T\t1287381\tlung_NSCLC\tlung_NSCLC_adenocarcinoma\tLUAD\tAmplification\tcnaLUAD27 (MYC)\n",
    "            22RV1\t924100\turogenital_system\tprostate\tPRAD\tDeletion\tcnaPRAD7\n",
    "\n",
    "    - Notebook: http://localhost:8912/notebooks/data_aggregation/raw/ccle/v1/gene-copy-number.ipynb\n",
    "    - How much should these tumors + genes overlap with cosmic?\n",
    "    - How should values be decoded?  See Appendix A\n",
    "\n",
    "\n",
    "\n",
    "- **Mutations**\n",
    "    - Downloaded spreadsheet: ~/repos/mgds/python/notebook/data_aggregation/raw/gdsc/v1/downloads/exome-seq.xlsx\n",
    "    - Notebook: http://localhost:8912/notebooks/data_aggregation/raw/gdsc/v1/gene-exome-seq.ipynb\n",
    "    - Any reason not to use \"Cell-line sequence variants\"?\n",
    "        - The legend for this is in Appendix B\n",
    "    - What is the \"Sequencing BEMs for Cell lines\" dataset? \n",
    "        - This downloads as a series of files named like:\n",
    "            - PANCAN_SEQ_BEM.txt\n",
    "            - SCLC_SEQ_BEM.txt\n",
    "            - BRCA_SEQ_BEM.txt\n",
    "        - each is csv with genes in rows and cell lines in columns, and 0s or 1s as values \n",
    "        - See /Users/eczech/Downloads/CellLines_CG_BEMs/LUAD_SEQ_BEM.txt for an example\n",
    "\n",
    "\n",
    "- **Drug Sensitivity**\n",
    "    - What to do with cases where there are two COSMIC drug ids for the same drug name?\n",
    "        - See plot in http://localhost:8912/notebooks/data_aggregation/import/gdsc/v2/drug-sensitivity.ipynb\n",
    "            - Dupes are correlated but not perfectly\n",
    "\n",
    "## NCI-60\n",
    "\n",
    "- **Drug Sensitivity**\n",
    "    - http://localhost:8912/notebooks/data_aggregation/import/nci60/v2/drug-sensitivity.ipynb\n",
    "    - What are \"probes\" in this context?\n",
    "    - What is QC process here, and should everything be ignored when num probes after QC is 0?\n",
    "    - How to deal with duplicates by drug name?  See list in notebook and examples with plots on similarity of GI50 across cell lines\n",
    "    \n",
    "    \n",
    "## CTD\n",
    "- **Drug Sensitivity**\n",
    "    - What does it mean for AUC to be high and EC50 to be very low?  It seems like this isn't altogether unexpected since AUC and IC50 measure different characteristics of the potency, but what does this say about the efficacy of the drug as a treatment?\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr>\n",
    "# Appendix A\n",
    "\n",
    "## GDSC Copy Number XLS Legend\n",
    "\n",
    "(downloaded from ftp://ftp.sanger.ac.uk/pub/project/cancerrxgene/releases/release-6.0/Gene_level_CN.xlsx)\n",
    "\n",
    "Gene level copy number analysis based PICNIC analysis of Affymetrix SNP6.0 array data.\n",
    "\n",
    "Four comma seperated peices of data are presented for each gene cell line combination (n1,n2,n3,n4), hyphen (-) is used where the value is unknown:-\n",
    "\n",
    "n1 - Maximum copy number of any genomic segment containing coding sequence of the gene (-1 indicates a value could not be assigned).\n",
    "\n",
    "n2 - Minimum copy number of any genomic segment containing coding sequence of the gene (-1 indicates a value could not be assigned).\n",
    "\n",
    "n3 - Zygosity - (H) if all segments containing gene sequence are heterozygous, (L) if any segment containing coding sequence has LOH, (0) if the complete coding sequence of the gene falls within a homozygous deletion.\n",
    "\n",
    "n4 - Disruption (D) if the gene spans more than 1 genomic segment (-) if no disruption occures.\n",
    "\n",
    "Therefore:\n",
    "\n",
    "2,2,H,-\tGene resides on a single genomic segment in a diploid region of the genome.\n",
    "\n",
    "2,0,L,D\tGene spans multiple segments, higest copy number is 2 but part of the coding sequence is homozygously deleted, the gene is disrupted.\n",
    "\n",
    "13,13,H,-\tGene resides on a single genomic segment of copy number 13 in a heterozygous part of the genome (amplification).\n",
    "\n",
    "14,12,L,D\tGene spans multiple genomic segments all of which are amplified to 12 or more copies, some or all segments have LOH, the gene is disrupted.\n",
    "\n",
    "0,0,0,-\tComplete gene sequence falls within a homozygous deletion."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr>\n",
    "# Appendix B\n",
    "\n",
    "(downloaded from ftp://ftp.sanger.ac.uk/pub/project/cancerrxgene/releases/release-6.0/WES_variants.xlsx)\n",
    "\n",
    "SAMPLE - Sample identifier provided in study\n",
    "\n",
    "COSMIC_ID - Unique numerical identifier for the cell lines used in COSMIC\n",
    "\n",
    "Cancer Type - TCGA tissue classification\n",
    "\n",
    "Gene - Gene name from Ensembl version 56\n",
    "\n",
    "Transcript - Transcript identifier from Ensembl version 56\n",
    "\n",
    "cDNA - Variant position and nucleotide change relating to the cDNA\n",
    "\n",
    "AA - Amno acid positon and alteration\n",
    "\n",
    "Classification - Summary of variant type\n",
    "\n",
    "Gene List - Identifier for variants within the 470 genes used in the study\n",
    "\n",
    "Recurrence Filter - Based on frequency observed in COSMIC (v68) (Recurrance filter: See Extended Experimental Procedures ) plus fusion gene data\n",
    "\n",
    "Subs - Missense/substitution variants occurring in codons mutated  in the systematic screen data in COSMIC (v68) (select >=3)\n",
    "\n",
    "Truncating - Truncating variant count  from the systematic screen data in COSMIC (v68) (select >10)\n",
    "\n",
    "inframe\t- Inframe indel alterations occurring in codons mutated in the systematic screen data in COSMIC (v68) (select >=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
